Better handling of OpInfo.sample_inputs

antocuni opened the pull request:
(@mruberry this is more or less what we discussed during a call ~1 month ago. Sorry if it took so long, in the meantime I had some real-life troubles which prevented me to work)

This fixes #55152 by implementing what I described as "option 2" in the issue. In summary:

1. Introduce `SampleInputCtx`: this is a context manager which wraps a `SampleInput`, intercept exceptions and add the name of the current sample input to the error message
2. `OpInfo.sample_inputs` now returns a list of `SampleInputCtx`
3. all the tests which used to do:
```python
    for sample in op.sample_inputs(...):
        ...
```
now do:
```python
    for ctx in op.sample_inputs(...):
        with ctx:
            ...
```
4. Introduce a new method `OpInfo.get_one_sample_input`: this simplifies a lot all the test which uses only the first `SampleInput` instead of iterating on all of them.

**Note for reviewers**: in some places the "textual diff" is larger than the "logical diff" because there are big chunks of text indented/dedented. I some some cases I have tried to minimize the diffs by doing the big indents/dedents in their own commits, so reviewing commits 1 by 1 might be easier than seeing the whole diff at once.

As a usage example, I have added reasonable names to the sample inputs of `linalg.det`. The tracebacks look like this:
```python traceback
Traceback (most recent call last):
  File "/work/antocuni/git/Quansight/pytorch/test/test_ops.py", line 60, in test_supported_dtypes
    raise Exception('fake error to test the traceback')
Exception: [SampleInput: SxS] fake error to test the traceback
```

If a `SampleInput` is not given an explicit name, `OpInfo.sample_inputs` automatically add one by using the index: this should be good enough to find easily find the culprit sample in most cases:
```python traceback
Traceback (most recent call last):
  File "/work/antocuni/git/Quansight/pytorch/test/test_ops.py", line 60, in test_supported_dtypes
    raise Exception('fake error to test the traceback')
Exception: [SampleInput: 0] fake error to test the traceback
```

If this PR gets merged, I'll be happy to submit subsequent PRs to add descriptive sample names where appropriate. E.g.
there are already other `sample_input_*` functions which have "sample names" as comments, e.g. `sample_inputs_einsum`:
https://github.com/pytorch/pytorch/blob/821a97595b3b0f8b1777cdd90a639078249c1846/torch/testing/_internal/common_methods_invocations.py#L2679-L2693

antocuni committed:
WIP: Introduce SampleInputCtx.

See the docstring for details, but the idea is that by wrapping
SampleInputs inside a context manager, we can intercept exceptions and
report the name of a specific SampleInput, to make it easier to debug.

antocuni committed:
attach names to SampleInputs for linalg_det

antocuni committed:
introduce OpInfo.get_one_sample_input and use it to simplify these three tests

antocuni committed:
fix test_sparse.py

mruberry commented on test/test_fx.py:
```diff
@@ -2595,23 +2595,24 @@ def test_get_torch_func_signature_exhaustive(self, device, dtype, op):
-                        pass
-                else:
-                    raise RuntimeError(f'Did not match any schemas for op {op.name}!')
+            for ctx in sample_ctxs:
```
fyi @Chillee @jamesr66a

mruberry replied:
I've suggested not updating fx code in this PR - so these changes will probably be mooted -- never mind the fyi

antocuni replied:
if I revert these changes and we keep the `ctx` API, these tests will start to fail. Do you mean that I should just do that?

mruberry commented on test/test_ops.py:
```diff
@@ -37,12 +37,10 @@ def test_unsupported_dtypes(self, device, dtype, op):
             # NOTE: only tests on first sample
-            sample = samples[0]
+            sample = op.get_one_sample_input(device, dtype)
+            if sample is None:
```
We should probably just remove this check: I think every current OpInfo has at least one SampleInput and every new OpInfo should have at least one

mruberry commented on test/test_ops.py:
```diff
@@ -52,12 +50,13 @@ def test_unsupported_dtypes(self, device, dtype, op):
-            self.assertTrue(sample_input.dtype == dtype)
-            self.assertTrue(sample_input.device.type == self.device_type)
+        for ctx in op.sample_inputs(device, dtype):
+            with ctx as sample:
```
An extra indentation is a pain, is there some pattern we could use to remove it?

mruberry replied:
Maybe something from contextlib https://docs.python.org/3/library/contextlib.html? Or setting the global exception handler? Or maybe there's a class-level exception handler that works with unittest and pytest?

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -78,17 +78,52 @@ def __init__(self, cls_name=None, test_name=None, *,
+
+    def __exit__(self, etype, evalue, tb):
+        if etype:
+            evalue.args = (f'[SampleInput: {self._sample.name}] {evalue}',)
```
What if the sample input doesn't have a name?

antocuni replied:
it shouldn't happen:
1. `SampleInput.__init__` sets `self.name` to `None` if none is given, so the line above never crashes
2. but most importantly, `OpInfo.sample_inputs()` automatically assigns a `name` based on the index, in case the name was none. So well-behaved code which uses the official API will always get named `SampleInput`s.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -303,6 +338,32 @@ def sample_inputs(self, device, dtype, requires_grad=False, **kwargs):
+        samples = self.generate_sample_inputs(device, dtype, requires_grad, **kwargs)
+        # make sure that every SampleInput has a sensible name: if the name
+        # was not given, use the index instead
+        for i, sample in enumerate(samples):
```
This is going to interfere with plans to make sample inputs use generators.

Why return the ctx and, if we take the context manager approach, not require tests just do:

```
with SampleGuard(sample):
```

for example? Seems like this complicates the process.

antocuni replied:
> This is going to interfere with plans to make sample inputs use generators.

I'm not sure to understand how. We can still `enumerate()` even if `samples` is a generator. And if for some reason we really want `OpInfo.sample_inputs()` to return a generator instead of a list, we can just put a `yield` in the for loop

> Why return the ctx and, if we take the context manager approach, not require tests just do:
```
with SampleGuard(sample):
```
> for example? Seems like this complicates the process.

my idea is that by returning the context, users are *forced* to use the `with` statement, and so we are sure that exceptions will always be intercepted by the guard. In your proposal, what would happen if I forget the `with SampleGuard(sample)`? Can I still use `sample` even outside the `with`?

mruberry replied:
Given that there are so few tests which use sample inputs I don't think we need to be very concerned with people forgetting to set the guard

mruberry commented on test/test_ops.py:
```diff
@@ -37,12 +37,10 @@ def test_unsupported_dtypes(self, device, dtype, op):
-
             # NOTE: only tests on first sample
-            sample = samples[0]
+            sample = op.get_one_sample_input(device, dtype)
```
For code velocity let's limit the application of this technique to this test for now (if we end up with a pattern where tests have to change at all), that way all the tests won't need to change and we can focus on the design

antocuni replied:
to get a good feedback of how the API will look like, we need at least a test which uses `get_one_sample_input` pattern and one which uses the `for sample in samples:` pattern. But indeed, it sounds like a good idea, I'll do it

mruberry reviewed (commented):
Hey @antocuni!

This is neat but I think we should look at a few possible patterns and continue supporting SampleInputs just being being SampleInputs, instead of returning context managers themselves. Maybe something from contextlib, for example, or being clever with the exception handlers would allow code to look like:

```
for sample in sample_inputs():
  self.sample_context(sample)
  ...
```

Which would be less disruptive. Or maybe the setting of the "sample context" could even be done with the SampleInput generator?

Looking forward to hearing your thoughts!

antocuni commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -445,34 +506,38 @@ def sample_inputs_tensor_split(op_info, device, dtype, requires_grad, **kwargs):
-        random_symmetric_psd_matrix(S, **kw),  # symmetric_psd
-        random_symmetric_pd_matrix(S, **kw),  # symmetric_pd
+    samples = [
+        sample('SxS', make_tensor((S, S), **kw)),
```
+1, it sounds like a good idea once pretty printing is available. What is the relevant PR?

mruberry replied:
@kshitij12345 Would you link it here?

antocuni commented:
> Hey @antocuni!
> 
> This is neat but I think we should look at a few possible patterns and continue supporting SampleInputs just being being SampleInputs, instead of returning context managers themselves. 

I agree that this would be nice to have. I even tried a couple of alternative approaches (such as using generators, using a `finally` and/or intercepting `StopIteration` to know "when the for loop is over" and similar) to get the "automatic context manager" behavior, but I couldn't find any which would work generally.

In particular, I tried to meet these two goals:

1. it should be impossible to "forget" the guard. I.e., if we are handling a `SampleInput`, it should always be reported
2. it should not give false negatives. I.e., if we are NOT handling a `SampleInput`, it should not be mistakenly reported.

> Maybe something from contextlib, for example, or being clever with the exception handlers would allow code to look like:
> 
> ```
> for sample in sample_inputs():
>   self.sample_context(sample)
>   ...
> ```

this is suboptimal because it's too easy to forget to call `self.sample_context`. Moreover, if we have additional asserts *after* the for loop, they would be mistakenly reported as belonging to the last sample.

> Which would be less disruptive. Or maybe the setting of the "sample context" could even be done with the SampleInput generator?

This idea looks more promising. And I think we can also avoid the context manager by intercepting the exceptions inside the function `instantiated_test` in `common_device_type.py`. I'll try to play with this idea and see how it looks like, thanks for the input!

antocuni committed:
WIP: refactor the whole approach around OpInfo.sample_inputs()

Instead of using the ctx approch, use a less invasive approach which
automatically do all of the work

antocuni commented:
@mruberry I implemented a new approach in 149c167585, which I think covers ~95% of our needs:
1. `OpInfo.sample_inputs` keeps track of the currently-yielded sample
2. the exceptions are automatically intercepted and decorated by `instantiated_test`

This way, most of the end-user code doesn't even need to change.

There are two use cases which I can think of in which this approach won't work:
1. if we preemptively collect all the samples outside a loop, e.g. `samples = list(op.sample_inputs())`, but I don't see any reason to do that
2. if we try to iterate over `op.sample_inputs()` while another iteration is in-progress. Again, I can't imagine any use case for this, but I added a check just to fail early in that case

I modified back `test_supported_dtypes` to use the new approach and it seems to work well. I look forward to hear your input!

mruberry commented:
> @mruberry I implemented a new approach in [149c167](https://github.com/pytorch/pytorch/commit/149c16758585057e4494d337f460c2af03a336be), which I think covers ~95% of our needs:
> 
> 1. `OpInfo.sample_inputs` keeps track of the currently-yielded sample
> 2. the exceptions are automatically intercepted and decorated by `instantiated_test`
> 
> This way, most of the end-user code doesn't even need to change.
> 
> There are two use cases which I can think of in which this approach won't work:
> 
> 1. if we preemptively collect all the samples outside a loop, e.g. `samples = list(op.sample_inputs())`, but I don't see any reason to do that
> 2. if we try to iterate over `op.sample_inputs()` while another iteration is in-progress. Again, I can't imagine any use case for this, but I added a check just to fail early in that case
> 
> I modified back `test_supported_dtypes` to use the new approach and it seems to work well. I look forward to hear your input!

That sounds awesome!

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -303,6 +309,52 @@ def sample_inputs(self, device, dtype, requires_grad=False, **kwargs):
+            assert isinstance(sample, SampleInput)
+            # make sure that every SampleInput has a sensible name: if the
+            # name was not given, use the index instead
+            if sample.name is None:
```
Using @kshitij12345's work I'd ditch the enumerate here and just print the sample input, named or unnamed

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -303,6 +309,52 @@ def sample_inputs(self, device, dtype, requires_grad=False, **kwargs):
+            yield
+        except Exception as exc:
+            if self._current_sample:
+                exc.args = (f'[SampleInput: {self._current_sample.name}] {exc}',)
```
With @kshitij12345's PR I think this can just print the sample input's summary

mruberry replied:
Found it! This one: https://github.com/pytorch/pytorch/pull/58295

I'll try and merge it soon

antocuni replied:
see https://github.com/pytorch/pytorch/pull/58481#issuecomment-850537695 for a detailed answer

mruberry commented on test/test_ops.py:
```diff
@@ -58,6 +56,7 @@ def test_supported_dtypes(self, device, dtype, op):
             sample_input = sample.input[0] if is_iterable_of_tensors(sample.input) else sample.input
             self.assertTrue(sample_input.dtype == dtype)
             self.assertTrue(sample_input.device.type == self.device_type)
+            #raise Exception('fake error to test the traceback')
```
debug code?

mruberry reviewed (commented):
I really like the guard but I really dislike the indent and wonder if we should do something even more general.

Instead of just focusing on SampleInputs, a more general problem is: "I have some data that I would really like to append to a failing test message - how can I do that if I have a ton of assertEquals and/or want to preserve the original failure message?" And it turns out that's a little tricky to do even if we assume we're using unittest or pytest. cc @pmeier who knows more about pytest and may be able to correct me. 

So what if we generalize the guard to act as this more generic mechanism? Could we do something like...

```
tl = TestLogger()
for sample in samples:
  tl.log(sample.summary())
```

I also wouldn't worry about requiring tests that use sample inputs do this because there are so few of those tests we can easily correct tests that accidentally don't do this.

pmeier commented:
> Instead of just focusing on SampleInputs, a more general problem is: "I have some data that I would really like to append to a failing test message - how can I do that if I have a ton of assertEquals and/or want to preserve the original failure message?" And it turns out that's a little tricky to do even if we assume we're using unittest or pytest.

Regardless of `unittest` or `pytest` I think this always requires some custom code. `torch.testing.assert_close` supports passing a callable for the `msg` parameter that could be used in this case. See https://github.com/pytorch/pytorch/pull/55890#issue-614290734 for an example. Furthermore, I proposed to expose our function to generate the message in https://github.com/pytorch/pytorch/issues/58383#issue-893160617, which would make adding information easier.

If that is not an option this could be a workaround:

```python
import contextlib
from typing import Iterator


@contextlib.contextmanager
def amend_error_msg(msg_fmtstr: str) -> Iterator[None]:
    try:
        yield
    except Exception as error:
        raise type(error)(msg_fmtstr.format(msg=str(error))) from error
```

Example:

```python
a = 0
b = 1

with amend_error_msg(f"{{msg}}\n\nFailure happened for a={a} and b={b}"):
    assert a == b, "Values don't match!"
```

```
AssertionError: Values don't match!

Failure happened for a=0 and b=1
```

This is similar to what we do internally to append offending indices or keys to the error message in case a container is checked:

https://github.com/pytorch/pytorch/blob/be4ba29d49566a1d9069b8fb35d4b9b44bb1f1c5/torch/testing/_asserts.py#L578

https://github.com/pytorch/pytorch/blob/be4ba29d49566a1d9069b8fb35d4b9b44bb1f1c5/torch/testing/_asserts.py#L602

antocuni commented:
> I really like the guard but I really dislike the indent and wonder if we should do something even more general.

I'm confused. The approach which I described in https://github.com/pytorch/pytorch/pull/58481#issuecomment-847152448 does **not** require an indent, and there is no need for explicit guards.

> Instead of just focusing on SampleInputs, a more general problem is: "I have some data that I would really like to append to a failing test message - how can I do that if I have a ton of assertEquals and/or want to preserve the original failure message?" And it turns out that's a little tricky to do even if we assume we're using unittest or pytest. cc @pmeier who knows more about pytest and may be able to correct me.
> 
> So what if we generalize the guard to act as this more generic mechanism? Could we do something like...
> 
> ```
> tl = TestLogger()
> for sample in samples:
>   tl.log(sample.summary())
> ```

a general mechanism to attach data to exceptions sounds like a good idea. Probably in addition to `tl.log` it should also have a key/value mechanism so that we can override a specific key multiple times during the course of a test. E.g., I'm thinking about something like this:

```python
class OpInfo:
    def sample_inputs(self):
        for sample in self.generate_sample_inputs:
            tl = magically_get_test_logger()
            tl.set('SampleInput', sample.summary())
            yield sample
        tl.remove('SampleInput') # or maybe tl.set('SampleInput', None) ?

# common_device_type.py

def instantiated_test(...):
    tl = magically_get_test_logger()
    try:
        test_fn() # execute the instantiated test function
    except Exception as e:
        tl.amend_error_message(e)
```

I think that this proposed solution takes the best of all the proposals we got so far:
1. it introduces a general mechanism which we can use everywhere to provide additional info for failing tests
2. it solves the concrete problem of SampleInput
3. it doesn't require any change in the current tests and doesn't introduce any extra indentation levels

The only trick point is how to implement `magically_get_test_logger()`: it is basically a "per-test global" which potentially needs to accessed from very different places:
- from the sw engineering point of view, the best solution is to explicitly pass it around. The main drawback is that depending on where it is needed, it might be complicated to pass down layers and layers of code. `OpInfo.sample_inputs` is a perfect example in this sense.

- from a more practical point of view, having the notion of `global active TestLogger` is the easiest thing to do

antocuni committed:
remove leftover of the ctx approach

mruberry commented:
> > I really like the guard but I really dislike the indent and wonder if we should do something even more general.
> 
> I'm confused. The approach which I described in [#58481 (comment)](https://github.com/pytorch/pytorch/pull/58481#issuecomment-847152448) does **not** require an indent, and there is no need for explicit guards.

You're right, of course. I must have been looking at the wrong commit. Yes that approach seems like it would work well.


> a general mechanism to attach data to exceptions sounds like a good idea. Probably in addition to `tl.log` it should also have a key/value mechanism so that we can override a specific key multiple times during the course of a test. E.g., I'm thinking about something like this:
> 
> ```python
> class OpInfo:
>     def sample_inputs(self):
>         for sample in self.generate_sample_inputs:
>             tl = magically_get_test_logger()
>             tl.set('SampleInput', sample.summary())
>             yield sample
>         tl.remove('SampleInput') # or maybe tl.set('SampleInput', None) ?
> 
> # common_device_type.py
> 
> def instantiated_test(...):
>     tl = magically_get_test_logger()
>     try:
>         test_fn() # execute the instantiated test function
>     except Exception as e:
>         tl.amend_error_message(e)
> ```
> 
> I think that this proposed solution takes the best of all the proposals we got so far:
> 
> 1. it introduces a general mechanism which we can use everywhere to provide additional info for failing tests
> 2. it solves the concrete problem of SampleInput
> 3. it doesn't require any change in the current tests and doesn't introduce any extra indentation levels
> 
> The only trick point is how to implement `magically_get_test_logger()`: it is basically a "per-test global" which potentially needs to accessed from very different places:
> 
> * from the sw engineering point of view, the best solution is to explicitly pass it around. The main drawback is that depending on where it is needed, it might be complicated to pass down layers and layers of code. `OpInfo.sample_inputs` is a perfect example in this sense.
> * from a more practical point of view, having the notion of `global active TestLogger` is the easiest thing to do

Maybe we should just start with the sample input handler you've proposed. We can return to the idea of a more general test logger later.

antocuni committed:
Merge remote-tracking branch 'upstream/master' into antocuni/sample-input-better-error

antocuni committed:
the default value of sample.name is now '', use a more general check for 'name-not-given'

antocuni committed:
OpInfo.sample_inputs MUST return at least 1 sample. Add a couple of sanity checks and remove the no-longer-needed skips in the tests

antocuni committed:
ensure to reset _current_sample in case of exception, else the subsequent calls to sample_inputs will fail

antocuni committed:
fix for tests which does not uses op

antocuni committed:
contextlib.nullcontext is 3.7+ only, introduce our own version

antocuni committed:
use a more robust approach to ensure that we don't carry OpInfo state between tests

antocuni committed:
fix sample_inputs_foreach to return a list SampleInputs

mruberry commented:
> I implemented the proposed approach, and now we get tracebacks with SampleInput info automatically in every test.
> (At the moment of writing tests are still running but I hope to have fixed all the remaining issues by now).
> 
> One thing still open to discussion is how to integrate my work with @kshitij12345's `SampleInput.summary()`.
> Currently, the tracebacks produced by my PR look like this (let's call it option 1):
> 
> ```
> Traceback (most recent call last):
>   File "/home/antocuni/git/Quansight/pytorch/test/test_ops.py", line 57, in test_supported_dtypes
>     raise Exception('fake error')
> Exception: [SampleInput: SxS] fake error
> ```
> 
> If I try to print `sample.summary()` instead of just the name, the tracebacks look like this (option 2):
> 
> ```
> Traceback (most recent call last):
>   File "/home/antocuni/git/Quansight/pytorch/test/test_ops.py", line 57, in test_supported_dtypes
>     raise Exception('fake error')
> Exception: [SampleInput(input=Tensor[5, 5], args=(), kwargs={}, output_process_fn_grad=None, broadcasts_input=False, name='SxS')] fake error
> ```
> 
> Let me propose a couple of additional options:
> 
> (option 3): similar to option 2 but putting the name first:
> 
> ```
> Traceback (most recent call last):
>   File "/home/antocuni/git/Quansight/pytorch/test/test_ops.py", line 57, in test_supported_dtypes
>     raise Exception('fake error')
> Exception: [SampleInput(name='SxS', input=Tensor[5, 5], args=(), kwargs={}, output_process_fn_grad=None, broadcasts_input=False)] fake error
> ```
> 
> (option 4.1): like option 3, but print only name and input:
> 
> ```
> Traceback (most recent call last):
>   File "/home/antocuni/git/Quansight/pytorch/test/test_ops.py", line 57, in test_supported_dtypes
>     raise Exception('fake error')
> Exception: [SampleInput(name='SxS', input=Tensor[5, 5])] fake error
> ```
> 
> (option 4.2): same as 4.1, but printing also the other parameters _if they have a non-default value_ (in this example, the traceback looks the same)
> 
> (option 5): similar to 4.1/4.2, but using a more human-friendly syntax instead of a python-like syntax which is harder to read:
> 
> ```
> Traceback (most recent call last):
>   File "/home/antocuni/git/Quansight/pytorch/test/test_ops.py", line 57, in test_supported_dtypes
>     raise Exception('fake error')
> Exception: [SampleInput: SxS, input=Tensor[5, 5]] fake error
> ```
> 
> Personally, I like option 5 but it's just a preference so I'm fine with whatever we decide.
> 
> Moreover, depending on what we decide, we should consider whether to simplify the following error message introduced by #58295, since the info about the SampleInput are already included in the traceback:
> https://github.com/pytorch/pytorch/blob/41054f2ab5bb39d28a3eb8497f1a65b42385a996/test/test_ops.py#L323-L327
> 
> Currently I get this traceback:
> 
> ```
> Traceback (most recent call last):
> ...
> AssertionError: [SampleInput: 6] RuntimeError not raised : inplace variant either incorrectly allowed resizing or you have marked the sample SampleInput(input=Tensor[5, 5], args=TensorList[Tensor[5, 5, 5]], kwargs={}, output_process_fn_grad=None, broadcasts_input=True, name='6') incorrectly with `broadcasts_self=True
> ```
> 
> If we use e.g. option 5 above and simplify the error message, we could get something like this:
> 
> ```
> AssertionError: [SampleInput: 6, input=Tensor[5, 5], args=TensorList[Tensor[5, 5, 5]]] RuntimeError not raised : inplace variant either incorrectly allowed resizing or you have marked the sample incorrectly with `broadcasts_self=True
> ```
> 
> @mruberry @kshitij12345 I'd love to hear your thoughts!

I like the simplest solution of just moving the name first. What do you think, @kshitij12345?

Good catch on that error message, @antocuni. Seems like we could fix it by just not printing the sample's summary when it's thrown?

kshitij12345 commented:
If this PR ascertains that each sample will have a name/serial number, then I think just the **name** is enough to provide pointer to the failing sample for further inspection.

antocuni commented:
> If this PR ascertains that each sample will have a name/serial number, then I think just the **name** is enough to provide pointer to the failing sample for further inspection.

yes, that's the case. If you don't give an explicit name, `OpInfo.sample_inputs()` will take care of assigning the serial number

antocuni committed:
use OpInfo.summary() to decorate the exceptions, and move the name in front to make it more visible

antocuni committed:
there is no longer need to explicitly print the sample since it's already automatically included in the tracebacks

antocuni committed:
fix this test

antocuni commented:
@mruberry:
1. I tweaked `OpInfo.summary()` to move the name in front and I'm just using it to decorate the exceptions now (see 31e2467b44)
2. I simplified `test_variant_consistency_eager` (see 5dc34ea571)

The traceback now looks like this:
```
Traceback (most recent call last):
  File "/home/antocuni/git/Quansight/pytorch/test/test_ops.py", line 341, in test_variant_consistency_eager
    _test_consistency_helper(samples, variants)
  File "/home/antocuni/git/Quansight/pytorch/test/test_ops.py", line 327, in _test_consistency_helper
    pass
  File "/home/antocuni/.conda/envs/pytorch-cuda-dev/lib/python3.8/unittest/case.py", line 227, in __exit__
    self._raiseFailure("{} not raised".format(exc_name))
  File "/home/antocuni/.conda/envs/pytorch-cuda-dev/lib/python3.8/unittest/case.py", line 164, in _raiseFailure
    raise self.test_case.failureException(msg)
AssertionError: [SampleInput(name='6', input=Tensor[5, 5], args=TensorList[Tensor[5, 5, 5]], kwargs={}, output_process_fn_grad=None, broadcasts_input=True)] RuntimeError not raised : inplace variant either incorrectly allowed resizing or you have marked the sample incorrectly with `broadcasts_self=True
```

Assuming that I fixed all the remaining failures, I think that the PR is ready for another round of review

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -328,21 +341,67 @@ def get_operator_variant(self):
+                sample.name = str(i)
+            self._current_sample = sample
+            yield sample
+        self._current_sample = None
```
I imagine this reference is why the tests are failing with CUDA memory leaks (because it's keeping the samples alive after the test's conclusion)

antocuni replied:
I think you are right, but it's an easy fix. I pushed d05a010730 which fixes the tests on my machine, let's see what happens on buildbots

mruberry reviewed (commented):
There are two things that are tricky about the current approach that make me wonder if we should pursue the more general logger:

- as @antocuni pointed out earlier, the current error system will print the last sample input acquired even if it's no longer relevant
- the current approach leaks CUDA memory

The more general logger design, where the logger is created in the test and then the append message is set on it, shouldn't have either of these issues, I think? It would also let us prototype the change for just one test and not require additional indentation, and it may be VERY useful outside of sample input testing. For example, users often ask for subtest support or the ability to parametrize, and providing a logger like this seems like it gives a lot of subtest or parametrize-like support.

Note that I wouldn't autocreate the logger but require each test that wanted to use it create one. That way it will be useful outside of OpInfo tests, too.

antocuni committed:
make sure to always reset a op at the END of each test: this way, we let tensors to be fully collected

antocuni committed:
Merge remote-tracking branch 'upstream/master' into antocuni/sample-input-better-error

antocuni committed:
fix merge artifact

antocuni committed:
fix flake8

antocuni commented:
> There are two things that are tricky about the current approach that make me wonder if we should pursue the more general logger:
> 
> * as @antocuni pointed out earlier, the current error system will print the last sample input acquired even if it's no longer relevant

not in the general case: `OpInfo.sample_inputs` reset `self._current_sample` to `None` at the end of the loop:
https://github.com/pytorch/pytorch/blob/33ce34aa457513f3ec740557e2b574194955c6b4/torch/testing/_internal/common_methods_invocations.py#L409-L417

What you describe can happen only if the generator is left pending, e.g. by code like this:
```python
samples = op.sample_inputs() # this is a generator
first_sample = next(sample)
# the generator is left pending and will be closed at the end of the function and/or by the GC
```

I am not aware of any code like this in the current codebase (especially because now you can use `op.get_one_sample_input()` to get the first sample).
Personally, I don't think it's a good enough reason to renounce to the advantages of this solution.

> * the current approach leaks CUDA memory

this is no longer the case, the leak was fixed by d05a010.

> The more general logger design, where the logger is created in the test and then the append message is set on it, shouldn't have either of these issues, I think? It would also let us prototype the change for just one test and not require additional indentation, and it may be VERY useful outside of sample input testing. 

The logger approach has some disadvantages as well, depending on how we implement it.
E.g., for the `OpInfo.sample_inputs()` usecase, you would need to explicitly pass the logger to `sample_inputs()` and/or somehow automatically inject the logger into the `OpInfo` instance.

Moreover, once we have a logger we also need a way to display the logged contents. The most reasonable thing to do is to collect all the logged info and display them on failure, but this is *exactly* what `pytest` and other test runners do with `print()`/`sys.stdout`. So, the first question to answer would be why such a logger would be better than `print()` and/or the standard `logging` system.

> For example, users often ask for subtest support or the ability to parametrize, and providing a logger like this seems like it gives a lot of subtest or parametrize-like support.
Note that I wouldn't autocreate the logger but require each test that wanted to use it create one. That way it will be useful outside of OpInfo tests, too.

I agree that the ability of having subtests and parametrize would be nice to have, but a custom logger would solve only a tiny bit of the problem. If we really want to support parametrization I would suggest to either migrate to `pytest` or to implement a 1st class parametrization system for our test suite, in the same style as the current `@ops` and `@dtypes` decorators. I would be glad to work on it if there is interest.

antocuni committed:
Merge remote-tracking branch 'upstream/master' into antocuni/sample-input-better-error

antocuni committed:
we can't do this check since it's now a generator, but also there should be no op with 0 samples nowadays

mruberry commented:
> The logger approach has some disadvantages as well, depending on how we implement it.
> E.g., for the `OpInfo.sample_inputs()` usecase, you would need to explicitly pass the logger to `sample_inputs()` and/or somehow automatically inject the logger into the `OpInfo` instance.

I would just have the test create it and write to the logger and not worry about combining the logic with sample inputs or OpInfos. 

> Moreover, once we have a logger we also need a way to display the logged contents. The most reasonable thing to do is to collect all the logged info and display them on failure, but this is _exactly_ what `pytest` and other test runners do with `print()`/`sys.stdout`. So, the first question to answer would be why such a logger would be better than `print()` and/or the standard `logging` system.

In our current setup printing from a test just prints from the test. The print output is not conditionally printed on the test failing.

> I agree that the ability of having subtests and parametrize would be nice to have, but a custom logger would solve only a tiny bit of the problem. If we really want to support parametrization I would suggest to either migrate to `pytest` or to implement a 1st class parametrization system for our test suite, in the same style as the current `@ops` and `@dtypes` decorators. I would be glad to work on it if there is interest.

See https://github.com/pytorch/pytorch/issues/52984 for the latest discussion on adopting pytest. People like it but there are definitely some tweaks and I think we'd need to investigate the requirements to switch better. We could implement our own @parametrize, which people do request. It wouldn't address this sample input issue, though. I think it would have to work just like pytest's parametrize functionality and I'm not sure how much of a pain that is to replicate or not.

antocuni commented:
> > The logger approach has some disadvantages as well, depending on how we implement it.
> > E.g., for the `OpInfo.sample_inputs()` usecase, you would need to explicitly pass the logger to `sample_inputs()` and/or somehow automatically inject the logger into the `OpInfo` instance.
> 
> I would just have the test create it and write to the logger and not worry about combining the logic with sample inputs or OpInfos.

so, basically you propose to do the following, if I understand it correctly?

```python
        for sample in op.sample_inputs(device, dtype, ...):
            test_logger.print('SampleInput: ', sample.name)
            ...
```
this would work of course, with two obvious drawbacks:
1. every test, especially the new ones, has to remember to log the sample's name
2. the name of the sample might be harder to find if there is a lot of normal and/or logged output

This is basically the opposite of what my original idea tried to implement, but as usual it's ultimately your call.

 
> In our current setup printing from a test just prints from the test. The print output is not conditionally printed on the test failing.

maybe it should? A test logger introduces a level of complexity which makes it harder to write and read test, and to inspect the cause of failures (when should I use the logger instead of a print? Do I need to pass the logger around? Why are my print() and log() printed out of order? etc...).

Maybe a pytest-like way of capturing the output and print only on failures might solve 99% of the problem?

> See #52984 for the latest discussion on adopting pytest. People like it but there are definitely some tweaks and I think we'd need to investigate the requirements to switch better. We could implement our own @parametrize, which people do request. It wouldn't address this sample input issue, though. I think it would have to work just like pytest's parametrize functionality and I'm not sure how much of a pain that is to replicate or not.

yes, the biggest problem of parametrizing sample inputs is that you would need to know in advance (during the collections/instantiation phase of tests) the number and the names of samples, which I seem to understand it's not always the case.

mruberry commented:
> so, basically you propose to do the following, if I understand it correctly?
> 
> ```python
>         for sample in op.sample_inputs(device, dtype, ...):
>             test_logger.print('SampleInput: ', sample.name)
>             ...
> ```
>
> this would work of course, with two obvious drawbacks:
> 
> 1. every test, especially the new ones, has to remember to log the sample's name
> 2. the name of the sample might be harder to find if there is a lot of normal and/or logged output
> 
> This is basically the opposite of what my original idea tried to implement, but as usual it's ultimately your call.

Giving tests control over what they log seems like a good first step. They could log the entire prettyprint of the sample input + additional data, even. In the future if this logging is an issue we could add a helper, like "get_samples" that accepts a logger and sets the each sample input for the test writer.


> > In our current setup printing from a test just prints from the test. The print output is not conditionally printed on the test failing.
> 
> maybe it should? A test logger introduces a level of complexity which makes it harder to write and read test, and to inspect the cause of failures (when should I use the logger instead of a print? Do I need to pass the logger around? Why are my print() and log() printed out of order? etc...).

We could think about changing how print works in the test suite, but while developing you sometimes really want a print to see what's going on.

It's true we do have to document any logger we develop, and people may complain that they don't understand the logger, but as long as it's opt-in and we have the documentation I don't have a lot of sympathy for that complaint because they could have read the docs before using it.

> Maybe a pytest-like way of capturing the output and print only on failures might solve 99% of the problem?

Maybe? I'm not familiar with it. 

> > See #52984 for the latest discussion on adopting pytest. People like it but there are definitely some tweaks and I think we'd need to investigate the requirements to switch better. We could implement our own @parametrize, which people do request. It wouldn't address this sample input issue, though. I think it would have to work just like pytest's parametrize functionality and I'm not sure how much of a pain that is to replicate or not.
> 
> yes, the biggest problem of parametrizing sample inputs is that you would need to know in advance (during the collections/instantiation phase of tests) the number and the names of samples, which I seem to understand it's not always the case.

Correct.

antocuni commented:
> We could think about changing how print works in the test suite, but while developing you sometimes really want a print to see what's going on.
> 
> > Maybe a pytest-like way of capturing the output and print only on failures might solve 99% of the problem?
> 
> Maybe? I'm not familiar with it.

by default, when running a test `pytest` captures its stdout and the stderr. The stdout/stderr are shown only if the tests fails.
If you need to see what's going on you can disable capturing by passing `--capture=no` or `-s` on the command line.

The stdout/stderr streams are captured using a method [similar to this](https://eli.thegreenplace.net/2015/redirecting-all-kinds-of-stdout-in-python/#id4), so they capture also the output written by C/C++ libraries.

antocuni committed:
simplify the approach. OpInfo.sample_inputs now automatically attaches a name to each sample, but it's not automatically shown in case of exception

antocuni committed:
add SampleInput.log() and start to use it in tests

antocuni commented:
@mruberry so, I simplified a lot the whole approach.
The current sample is no longer automatically reported in case of exception, but every test is supposed to call `sample.log()` to ensure that the current sample is correctly reported.

For now, `SampleInput.log()` does a simple `print()`, which is more than enough to solve the problem at hand.
I think that a proper logger should be discussed/designed elsewhere, and we can use it once we have.

For now, I demonstrated the used only in `test_unsupported_backward`: if you are fine with the approach, I'll rebase on top of the current master, fix the conflicts and add the relevant `sample.log()` everywhere.

mruberry commented:
Hey @antocuni! Sorry for taking so long to get back to you, we were on vacation here in the States.

> @mruberry so, I simplified a lot the whole approach.

Cool!

> The current sample is no longer automatically reported in case of exception, but every test is supposed to call `sample.log()` to ensure that the current sample is correctly reported.

That sounds reasonable.

> For now, `SampleInput.log()` does a simple `print()`, which is more than enough to solve the problem at hand.
> I think that a proper logger should be discussed/designed elsewhere, and we can use it once we have.

Does that mean that tests will print debug information even if they don't fail? I don't think we should do that. If anything I'd like to add verification that tests ONLY print if they fail.

> For now, I demonstrated the used only in `test_unsupported_backward`: if you are fine with the approach, I'll rebase on top of the current master, fix the conflicts and add the relevant `sample.log()` everywhere.

Awesome; I'll take a look now.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -146,6 +146,14 @@ def formatter(arg):
+        by tests, to make it possible to know which SampleInput was in use in
+        case of a test failure
+        """
+        print(f'[SampleInput] {self.name}')
```
I think printing only on failure is the killer feature we want to go for with this PR. 

See https://github.com/pytorch/pytorch/issues/60748, I think it'd be generally useful.

antocuni replied:
I tried to design a solution in PR #61910

antocuni committed:
Introduce TestLogger and use it to augment error messages in case of exception

antocuni committed:
kill leftover from a previous approach

antocuni committed:
run test_logger_reporting only on CPU, else it fails if the e.g. there are env variables which filter out CPU tests

antocuni committed:
Merge remote-tracking branch 'upstream/master' into antocuni/sample-input-better-error

antocuni committed:
fix typo after merge

antocuni committed:
re-add an example of using sample.log()

antocuni committed:
fix

antocuni committed:
Merge branch 'antocuni/testlogger' into antocuni/sample-input-better-error

antocuni committed:
use the new TestLogger feature to log the sample names

antocuni committed:
test and fix for exceptions without a message

antocuni committed:
Merge branch 'antocuni/testlogger' into antocuni/sample-input-better-error

antocuni commented:
@mruberry I rebased this PR on top of #61910 (with the hope that it will be merged) so that I can use the new `TestLogger` to finally implement this feature.

I demonstrate the new approach in e.g. `test_grad`:

```python
        for sample in samples:
            sample.log(self.logger)
            if sample.broadcasts_input and is_inplace(variant):
                continue
        [...]
```

And this is the output I get if I introduce a failure when handling the sample of `linalg_det` called `symmetric`:
```
======================================================================
FAIL: test_fn_grad_linalg_det_cpu_float64 (__main__.TestGradientsCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/antocuni/git/Quansight/pytorch/torch/testing/_internal/common_device_type.py", line 376, in instantiated_test
    result = test(self, **param_kwargs)
  File "/home/antocuni/git/Quansight/pytorch/torch/testing/_internal/common_device_type.py", line 818, in dep_fn
    return fn(slf, *args, **kwargs)
  File "/home/antocuni/git/Quansight/pytorch/torch/testing/_internal/common_device_type.py", line 818, in dep_fn
    return fn(slf, *args, **kwargs)
  File "/home/antocuni/git/Quansight/pytorch/torch/testing/_internal/common_device_type.py", line 784, in test_wrapper
    return test(*args, **kwargs)
  File "test/test_ops.py", line 595, in test_fn_grad
    self._grad_test_helper(device, dtype, op, op.get_op())
  File "test/test_ops.py", line 580, in _grad_test_helper
    return self._check_helper(device, dtype, op, variant, 'gradcheck', check_forward_ad=check_forward_ad)
  File "test/test_ops.py", line 553, in _check_helper
    assert len(sample.name) < 5
AssertionError:
~~~~~~~~ TestLogger ~~~~~~~~
    [SampleInput] SxS
    [SampleInput] 1x1
    [SampleInput] symmetric
~~~~~~~~~~~~~~~~~~~~~~~~~~~~
````

Once we are happy with the approach, I'll add the various `sample.log()` in all the other places which call `op.sample_inputs()`

antocuni committed:
fix test_foreach

antocuni committed:
Merge remote-tracking branch 'upstream/master' into antocuni/testlogger

antocuni committed:
Merge branch 'antocuni/testlogger' into antocuni/sample-input-better-error

antocuni commented:
@mruberry I've rebased this again on top of #61910: once it lands, it should be safe to merge PR as well

mruberry commented:
> @mruberry I've rebased this again on top of #61910: once it lands, it should be safe to merge PR as well

Thanks @antocuni but don't worry I can handle the rebase when it's the right time to fit this in
